### PR TITLE
PhpUnitTestAnnotationFixer: Only remove prefix if it really is a prefix

### PR DESCRIPTION
**Pull request by localheinz**
This PR

* [x] adds more test cases
* [x] modifies detection of test prefix
* [x] removes `test` prefix only if the next character is not a number, nor a lower-case letter
* [x] removes a few related comments

Follows #3275.

### CODE REVIEW

File: src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php [hunk 1]
```diff
@@ -250,6 +250,56 @@ private function isFunction(Tokens $tokens, $index)
...
+    /**
+     * @param string $functionName
+     *
+     * @return bool
+     */
+    private function canRemoveTestPrefix($functionName)
+    {
+        $length = strlen($functionName);
+
+        if ($this->startsWith('test_', $functionName) && $length > 5) {
```
Comment (BackEndTea) [inline src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php]: Maybe add a test case to prove a function just being named ```test```or ```test_``` does not get renamed either

Comment (localheinz) [reply src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php]: @BackEndTea 

Good point, will take a look later!

File: src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php [hunk 2]
```diff
@@ -291,26 +291,36 @@ private function getDockBlockIndex(Tokens $tokens, $index)
...
+     */
+    private function hasTestPrefix($functionName)
+    {
+        if (!$this->startsWith('test', $functionName) || strlen($functionName) < 5) {
+            return false;
+        }
+
+        $nextCharacter = $functionName[4];
+
+        return is_numeric($nextCharacter) || $nextCharacter === strtoupper($nextCharacter);
```
Comment (localheinz) [inline src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php]: The second part covers also cases where we have `test_` as prefix, as `'_' === strotoupper('_')`.

File: src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php [hunk 3]
```diff
@@ -291,6 +291,22 @@ private function getDockBlockIndex(Tokens $tokens, $index)
...
+     */
+    private function hasTestPrefix($functionName)
+    {
+        if (!$this->startsWith('test', $functionName) || strlen($functionName) < 5) {
+            return false;
+        }
+
+        $nextCharacter = $functionName[4];
+
+        return $nextCharacter === strtoupper($nextCharacter);
```
Comment (localheinz) [inline src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php]: If we preferred to be explicit here, then this should be

```php
return is_numeric($nextCharacter) || '_' === $nextCharacter || $nextCharacter === strtoupper($nextCharacter);
```

instead.

File: src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php [hunk 4]
```diff
@@ -291,26 +291,36 @@ private function getDockBlockIndex(Tokens $tokens, $index)
...
-        if ($this->startsWith('test_', $functionName)) {
-            if (is_numeric(substr($functionName, 6, 1))) {
-                return $functionName;
-            }
+        $remainder = preg_replace('/^test_?/', '', $functionName);
 
-            return  substr($functionName, 5);
-        }
-        if (is_numeric(substr($functionName, 5, 1))) {
+        if ('' === $remainder || is_numeric($remainder[0])) {
```
Comment (localheinz) [inline src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php]: Can't remove the prefix if the remainder is an empty string, or the first character is a digit.

**Comment by SpacePossum** [comment]
LGTM, sadly the test fails on a not related issue (`echo: write error: Broken pipe`)

**Comment by localheinz** [comment]
Thank you, @SpacePossum! Noticed this locally as well, but focused only on `PhpUnitTestAnnotation` fixer here.

**Comment by localheinz** [comment]
@SpacePossum 

Initially I had three methods

- `hasTestPrefix()`
- `canRemoveTestPrefix()`
- `removeTestPrefix()`

What do you think, would the code be easier to understand if there was a single method to handle all of that? Similarly, for adding a test prefix, if necessary? After all, there’s some temporal coupling involved, as we always invoke `hasTestPrefix()`, immediately followed by `removeTestPrefix()`.

**Comment by keradus** [comment]
then `removeTestPrefix` would need to check `hasTestPrefix` internally to decouple the methods.
if there is a need to call one method but not another (even conditionally), then 2 methods are fine.
if we always call them in chain (even conditionally), I don't see value of splitting them

**Comment by localheinz** [comment]
@keradus @SpacePossum 

On second thought, we use `hasTestPrefix()` in two different scenarios:

* when verifying that we need to remove
* when verifying that we need to add

So I guess it's ok to leave them separate.

How about a bit of renaming, though? Please hang on for the next commit.

File: src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php [hunk 5]
```diff
@@ -140,29 +140,29 @@ private function findPhpUnitClasses(Tokens $tokens)
...
             $functionNameIndex = $tokens->getNextMeaningfulToken($i);
             $functionName = $tokens[$functionNameIndex]->getContent();
 
             if ($this->hasTestPrefix($functionName)) {
-                $newFunctionName = $this->removeTestFromFunctionName($functionName);
+                $newFunctionName = $this->removeTestPrefix($functionName);
                 $tokens[$functionNameIndex] = new Token([T_STRING, $newFunctionName]);
             }
 
-            $docBlockIndex = $this->getDockBlockIndex($tokens, $i);
```
Comment (localheinz) [inline src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php]: Here's a small typo in the method name.

**Comment by SpacePossum** [comment]
the current PR'ed composition of methods and their naming looks spot on to me :+1:

**Comment by localheinz** [comment]
@BackEndTea @keradus @SpacePossum 

I added one more test case, but other than that, I can't think of anything else to add at the moment.

**Comment by BackEndTea** [comment]
Thanks for improving my code a bit @localheinz

**Comment by keradus** [comment]
Somehow I missed that.
@localheinz , please rebase so I could merge it

**Comment by localheinz** [comment]
@keradus 

Rebased!

**Comment by keradus** [comment]
Thank you @localheinz.

**Comment by localheinz** [comment]
Thank you, @BackEndTea, @keradus, and @SpacePossum!

**Pull request closed without merging**